Advanced Keycloak CR configuration (#11065)
keycloak/keycloak: docs/guides/src/main/operator/advanced-configuration.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/options.adoc" as opts>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Advanced configuration"
summary="How to tune advanced aspects of the Keycloak CR">

== Advanced Configuration

> In this guide we will show how advanced concepts and options provided by Custom Resources (CR).

Reviewer: Also, that sentence does not make sense for me bc. of the "how". how what? should that be "In this guide, we will show how to use advanced concepts and options provided by Custom Resources (CR)"? Another, imo better alternative: "In this guide, you'll learn how to configure your Keycloak deployment using advanced concepts and options provided by Custom Resources (CR)"

=== Server Configuration details

> The `serverConfiguration` field of the Keycloak CR allow to pass to Keycloak any available configuration in the form of key-value pairs.

Reviewer: ```suggestion
The `serverConfiguration` field of the Keycloak CR allows to pass to Keycloak any available configuration in the form of key-value pairs.
```

> For all the available configuration options refer to <@links.server id="all-config"/>.

Reviewer: ```suggestion
For all available configuration options, refer to <@links.server id="all-config"/>.
```

Reviewer: (incorporated @abstractj feedback into suggestion and removed "the")

The values can be expressed as plain text strings or Kubernetes Secret references.
e.g:

[source,yaml]
----
apiVersion: keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  ...
  serverConfiguration:
    - name: db
      value: postgres # plain text value
    - name: db-url-host
      value: postgres-db # plain text value
    - name: db-username
      secret: # Secret reference
        name: keycloak-db-secret # name of the Secret
        key: username # name of the Key in the Secret
    - name: db-password
      secret: # secret reference
        name: keycloak-db-secret # name of the Secret
        key: password # name of the Key in the Secret
----

=== Secret References

A Secret Reference can be either a value in `serverConfiguration` or the `tlsSecret`.

> When specifying a Secret Reference you have to make sure that a Secret containing the referenced keys is present in the same namespace as the CR referencing it.

Reviewer 2: "...a Secret Reference, you have..."

> Along with the Keycloak Server Deployment the operator will add special labels to the referenced Secrets in order to watch for changes.

Reviewer 2: "...Keycloak Server Deployment, the operator..."

> Along with the Keycloak Server Deployment the operator will add special labels to the referenced Secrets in order to watch for changes.

Reviewer: ```suggestion
Along with the Keycloak Server Deployment, the operator adds special labels to the referenced Secrets in order to watch for changes.
```

Reviewer: (changes from bruno incorporated in suggestion)

> When a referenced Secret is modified the operator will automatically perform a rolling restart of the Keycloak Deployment to pick up the changes.

Reviewer 2: "...a referenced Secret is modified, the operator..."

Reviewer: ```suggestion
When a referenced Secret is modified, the operator automatically performs a rolling restart of the Keycloak Deployment to pick up the changes.
```

Reviewer: (changes from bruno incorporated in suggestion)

=== Unsupported features

The `unsupported` field of the CR contains highly experimental configuration options that are not completely tested and supported.

==== Pod Template

Pod Template is a raw API representation that is used for the Kubernetes Deployment Template.

.Note:
[NOTE]

> You should use this field as a temporary workaround if there is no officially supported field at the top level of the CR to cover your use-case.

Reviewer: When writing the kc17 docs, we tried to make nearly no use of ".Note", as you could normally incorporate your "Note" into the normal text or give it its own sub-headline. Example for this one: 

"Pod Template is a raw API representation that is used for the Kubernetes Deployment Template. This field is intended to be used as a temporary workaround if there is no officially supported field at the top level of the CR to cover your use-case. Please consider opening an issue on GitHub to help us make the experience better."

That said: not hard on this ;) @stianst suggested to not make heavy use of notes, so I tried not to.

Author: I like this wording!

The operator will merge the fields of the provided template with the values generated by the operator for the specific Deployment.

> Using this feature you have access to a high level of customizations but there are no guarantees that the Deployment will work as expected.

Reviewer 2: "...high level of customizations, but there..."

> Using this feature you have access to a high level of customizations but there are no guarantees that the Deployment will work as expected.

Reviewer 2: "Using this feature, you..."

> As an example you can inject labels, annotations or even volumes and volume mounts, e.g:

Reviewer 2: "As an example, you can inject labels, annotations, or even volumes..."

Reviewer: I'd also remove the "e.g.", because sentence starts with "As an example". See suggestion (incorporating Brunos changes) below:

Reviewer: ```suggestion
As an example, you can inject labels, annotations, or even volumes and volume mounts:
```

[source,yaml]
----
apiVersion: keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  ...
  unsupported:
    podTemplate:
      metadata:
        labels:
          my-label: "keycloak"
      spec:
        containers:
          - volumeMounts:
              - name: test-volume
                mountPath: /mnt/test
        volumes:
          - name: test-volume
            secret:
              secretName: keycloak-additional-secret
----

=== Disabling required CR fields

> The Keycloak operator is designed to provide you, by default, with the best production ready Deployment of Keycloak with security in mind.

Reviewer 2: "...best production-ready..."

Reviewer: ```suggestion
The Keycloak operator is designed to provide you by default with the best production-ready Deployment of Keycloak with security in mind.
```

Reviewer: these commata felt weird. not native speaker though, so not sure ;) feel free to ignore suggestion. Also asking myself if there shouldn't be a comma at "of Keycloak, with security..."

> Although for development purposes you can still disable key security features.

Reviewer 2: "Although, for development purposes, you can..."

> Although for development purposes you can still disable key security features.

Reviewer 3: Could you please elaborate more on how disabling these features for development purposes are related to just running the server in dev mode using `start-dev`?

Wouldn't be easier to allow using CRs marked as `dev` so that pods are started in dev mode?

Author: Having the opportunity to granularly disable specific features instead of switching to a different profile has some advantages:
- you can disable a specific feature instead of opting in for a completely different configuration
- makes testing easier (and it's always a good sign 🙂  )
- disabling each specific feature is a deliberate and clearly discouraged practice the user would need to do

Last but not least, we want the operator to be biased toward production-ready installations and those options are offered as escape hatches for specific situations.

I do not consider the option of letting the user decide for the `dev` profile out of the table, but I believe this is the first step and we can iterate according to the community feedback after the preview version of this operation is out.

Does this answer your question @pedroigor ?

Reviewer 4: +1 on what @andreaTP said

> Spefically you can disable the required fields with a special value `INSECURE-DISABLE`, e.g:

Reviewer 2: Typo: Spefically

Specifically, you
